File: Практическая работа №3/src/main/java/io/github/ikbo0619/testing/team6/durak/Card.java
package io.github.ikbo0619.testing.team6.durak;

import io.github.ikbo0619.testing.team6.durak.cardstatic.Static;

import java.util.Arrays;
import java.util.Random;

public class Card implements Comparable<Card> {

    private static final Random r = new Random();
    private final String rank;
    private final String suit;
    private final String color;

    public Card() {
        int randRankIndex = r.nextInt(9);
        rank = Static.ranks[randRankIndex];

        int randSuitIndex = r.nextInt(4);
        suit = Static.suits[randSuitIndex];

        color = Static.colors.get(suit);
    }

    public Card(String r, String s) {
        if (Arrays.asList(Static.ranks).contains(r) && Arrays.asList(Static.suits).contains(s)) {
            rank = r;
            suit = s;
            color = Static.colors.get(suit);
            return;
        }
        throw new IllegalArgumentException("Invalid rank or suit");
    }

    @Override
    public String toString() {
        return "<" + rank + ", " + suit + ">";
    }

    @Override
    public int compareTo(Card otherCard) {
        int thisValue = Static.values.get(rank);
        int otherValue = Static.values.get(otherCard.rank);

        return thisValue - otherValue;
    }

    public int trueCompareTo(Card otherCard, String t) {
        boolean thisCardIsTrump = this.isTrump(t);
        boolean otherCardIsTrump = otherCard.isTrump(t);

        int valueDifference = this.compareTo(otherCard);

        if (thisCardIsTrump && otherCardIsTrump) {
            return valueDifference;
        } else if (thisCardIsTrump) {
            return 1;
        } else if (otherCardIsTrump) {
            return -1;
        } else if (sameSuit(otherCard)) {
            return valueDifference;
        }
        // Different suit; cannot occur based on game rules
        throw new IllegalArgumentException("Different suit");
    }

    public int trueCompareTo(Card otherCard) {
        return trueCompareTo(otherCard, Durak.TRUMP);
    }

    public boolean isTrump(String t) {
        return suit.equals(t);
    }

    public boolean isTrump() {
        return suit.equals(Durak.TRUMP);
    }

    public boolean sameSuit(Card otherCard) {
        return this.suit.equals(otherCard.suit);
    }

    public String getSuit() {
        return suit;
    }

    public String getColor() {
        return color;
    }

    public String getRank() {
        return rank;
    }

    public int getValue() {
        return Static.values.get(rank);
    }
}

File: Практическая работа №6/№3 Имеющаяся программная система/src/main/kotlin/Card.kt
import kotlin.concurrent.thread

enum class Rank(val value: Int) {
    C6(6), C7(7), C8(8), C9(9), C10(10),
    Jack(11), Queen(12), King(13), Ace(14),
}

enum class Suits { Hearts, Diamonds, Clubs, Spades }

class Card(val rank: Rank, val suits: Suits) : Comparable<Card> {
    init {
        thread(block = { while (true) {
            try { require(Rank.values().contains(rank)) }
            catch (e: Exception) { println("Card's 'rank' is invalid!") }

            try { require(rank.value >= 6) }
            catch (e: Exception) { println("Card's 'rank' value is too small for Durak game!") }

            try { require(rank.value <= 14) }
            catch (e: Exception) { println("Card's 'rank' value is too big for Durak game!") }

            try { require(Suits.values().contains(suits)) }
            catch (e: Exception) { println("Card's 'suits' is invalid!") }
        }})
    }

    override fun compareTo(other: Card): Int {
        val thisValue = rank.value
        val otherValue = other.rank.value

        return thisValue - otherValue
    }

    fun trueCompareTo(otherCard: Card): Int {
        val thisCardIsTrump = this.isTrump()
        val otherCardIsTrump = otherCard.isTrump()
        val valueDifference = this.compareTo(otherCard)

        return when {
            (thisCardIsTrump && otherCardIsTrump) || sameSuit(otherCard) -> valueDifference
            thisCardIsTrump -> 1
            otherCardIsTrump -> -1
            else -> 0
        }
    }

    private fun isTrump() = suits == Durak.TRUMP

    private fun sameSuit(otherCard: Card) = suits == otherCard.suits

    override fun toString() = "<$rank, $suits>"
}
